File: fabric-audit-agent/core/outcomes.js
/**
 * Compare the current run to the previous one: which active findings disappeared
 * (resolved/fixed), and how the peak-CU metric moved. Pure.
 * @param {object[]} currentFindings  current active findings (with .key)
 * @param {Array<{metrics?:{peakCuPct?:number}, findings?:Array<{key:string, suppressed?:boolean}>}>} history  chronological
 * @param {number|null} currentMetric  current peakCuPct
 * @returns {{ resolvedSinceLast: string[], metricDelta: object|null }}
 */
export function assessOutcomes(currentFindings = [], history = [], currentMetric = null) {
  if (!history.length) return { resolvedSinceLast: [], metricDelta: null };
  const prev = history[history.length - 1];
  const prevActive = new Set((prev.findings ?? []).filter(f => !f.suppressed && f.key).map(f => f.key));
  const cur = new Set(currentFindings.map(f => f.key));
  const resolvedSinceLast = [...prevActive].filter(k => !cur.has(k));

  let metricDelta = null;
  const from = prev.metrics?.peakCuPct;
  if (typeof from === 'number' && typeof currentMetric === 'number') {
    metricDelta = {
      metric: 'peakCuPct',
      from,
      to: currentMetric,
      change: Math.round((currentMetric - from) * 10) / 10,
      improved: currentMetric < from,
    };
  }
  return { resolvedSinceLast, metricDelta };
}

/** One-line human summary, or '' if nothing to report. Pure. */
export function summarizeOutcomes(outcomes = { resolvedSinceLast: [], metricDelta: null }) {
  const parts = [];
  if (outcomes.resolvedSinceLast.length) {
    parts.push(`${outcomes.resolvedSinceLast.length} finding(s) resolved since the last run`);
  }
  if (outcomes.metricDelta) {
    const d = outcomes.metricDelta;
    parts.push(`peak CU ${d.improved ? 'improved' : 'rose'} ${d.from}% → ${d.to}%`);
  }
  return parts.join('; ');
}

File: fabric-audit-agent-app/fabric_audit_agent/outcomes.py
"""Run-over-run outcomes: what got resolved + how the metric moved. Port of ``core/outcomes.js``. Pure."""
import math


def _is_num(v):
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _r1(x):
    return math.floor(x * 10 + 0.5) / 10


def assess_outcomes(current_findings=None, history=None, current_metric=None):
    current_findings = current_findings or []
    history = history or []
    if not history:
        return {"resolvedSinceLast": [], "metricDelta": None}
    prev = history[-1]
    # preserve prev.findings order (matches JS Set insertion order)
    prev_active, seen = [], set()
    for f in (prev.get("findings") or []):
        k = f.get("key")
        if not f.get("suppressed") and k and k not in seen:
            seen.add(k)
            prev_active.append(k)
    cur = {f.get("key") for f in current_findings}
    resolved_since_last = [k for k in prev_active if k not in cur]

    metric_delta = None
    frm = (prev.get("metrics") or {}).get("peakCuPct")
    if _is_num(frm) and _is_num(current_metric):
        metric_delta = {
            "metric": "peakCuPct", "from": frm, "to": current_metric,
            "change": _r1(current_metric - frm), "improved": current_metric < frm,
        }
    return {"resolvedSinceLast": resolved_since_last, "metricDelta": metric_delta}


def summarize_outcomes(outcomes=None):
    outcomes = outcomes or {"resolvedSinceLast": [], "metricDelta": None}
    parts = []
    if outcomes["resolvedSinceLast"]:
        parts.append(f"{len(outcomes['resolvedSinceLast'])} finding(s) resolved since the last run")
    if outcomes.get("metricDelta"):
        d = outcomes["metricDelta"]
        parts.append(f"peak CU {'improved' if d['improved'] else 'rose'} {d['from']}% → {d['to']}%")
    return "; ".join(parts)
